go/vt/vtctl/workflow/server.go

```diff
@@ -2139,27 +2158,46 @@ func (s *Server) WorkflowDelete(ctx context.Context, req *vtctldatapb.WorkflowDe
 		s.optimizeCopyStateTable(tablet.Tablet)
 		return res.Result, err
 	}
+
+	// Multi-tenant migrations delete only that tenant's records from the target tables
+	// in batches and we may not be able to complete that work before the timeout. We
+	// delete the workflow only after the cleanup work completes successfully so the
+	// workflow can be canceled multiple times if needed in order to fully cleanup
+	// all of the tenant's data that we had copied.
+	if ts.IsMultiTenantMigration() {
+		if ts.workflowType != binlogdatapb.VReplicationWorkflowType_MoveTables { // Should never happen
+			return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "unsupported workflow type %q for multi-tenant migration",
+				ts.workflowType)
+		}
+		// We need to delete the rows that the target tables would have for the tenant.
+		// We don't cleanup other related artifacts since they are not tied to the tenant.
+		if !req.GetKeepData() {
+			if err := s.DeleteTenantData(ctx, ts, req.DeleteBatchSize); err != nil {
+				return nil, vterrors.Wrapf(err, "failed to fully delete all migrated data for tenant %s, please retry the operation",
+					ts.options.TenantId)
+			}
+		}
+	} else {
+		// Cleanup related data and artifacts. There are none for a LookupVindex workflow.
+		if ts.workflowType != binlogdatapb.VReplicationWorkflowType_CreateLookupIndex {
+			if _, err := s.DropTargets(ctx, ts, req.GetKeepData(), req.GetKeepRoutingRules(), false); err != nil {
+				if topo.IsErrType(err, topo.NoNode) {
+					return nil, vterrors.Wrapf(err, "%s keyspace does not exist", req.GetKeyspace())
+				}
+				return nil, err
+			}
+		}
+	}
+
+	// Now that we've succesfully cleaned up everything else, we can finally delete
+	// the workflow.
 	delCtx, delCancel := context.WithTimeout(ctx, topo.RemoteOperationTimeout*2)
 	defer delCancel()
 	res, err := vx.CallbackContext(delCtx, callback)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(res) == 0 {
-		return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "the %s workflow does not exist in the %s keyspace", req.Workflow, req.Keyspace)
-	}
```

Comment:
Is this check no longer necessary? Or was it redundant in the first place?

Reply:
We now delete the actual workflow at the end, and by this time we've already read the workflow details. Followed by cleaning up all related artifacts. So at this point if the workflow itself is gone that's not an error condition.